text/3340-exhaustive-output.md

```diff
@@ -0,0 +1,111 @@
+- Feature Name: exhaustive_match_output
+- Start Date: 2022-10-31
+- RFC PR: [rust-lang/rfcs#3340](https://github.com/rust-lang/rfcs/pull/3340)
+- Rust Issue: [rust-lang/rust#103818](https://github.com/rust-lang/rust/issues/103818)
+
+# Summary
+[summary]: #summary
+
+Add an attribute which is only applicable to match statements called `#[exhaustive_output]` which will verify that
+the match statement has an arm for every variant of the enum type which it returns. This attribute will not be
+applicable to match statements that don't return an enum type.
+
+# Motivation
+[motivation]: #motivation
+
+Oftentimes it is necessary to convert unrestricted values into enum variants. Unrestricted values can include integers, strings, UUIDs, even other much larger enums.
```

Comment:
Note that integers *are* exhaustive -- we even use that in matching, where you don't need wildcards to match integers (or `char`s, for that matter!).

So shouldn't this use that same exhaustiveness checking?  What if I wrote
```
#[exhaustive_output]
match my_u8 {
    0 => 255,
    x @ 1..=254 => x,
    255 => 1, // Oops, I meant `0`
}
```
Could that give me a warning that I wasn't exhaustive?

More plausibly, if we get something like [`#[niche]`](https://github.com/rust-lang/rfcs/pull/3334), then arguably anything we can support for enums we should be supporting for niched integers too -- otherwise it's arguably better to `#[repr(u8)] enum Foo { _1 = 1, _2 = 2, _3 = 3 }` than to use `RefinedU8<1..=3>`.

Reply:
I suppose so, though from an implementation perspective this complicates things a lot as the compiler has to infer the possible outputs of `x @ 1..=254 => x,`, and pretty much every other integer pattern.